### DISCUSSION TITLE
Re-Use and Save Model Object

### DISCUSSION THREAD

**Discussion by Rjacobucci** [cdriveraus/ctsem#42] [General]
First off -- thanks for the great package! 

I am hoping to get thoughts on the best way to "train" a model, save the model object, which can then be used with new participants to update the parameters. Right now I'm having difficulty in running a model, saving the object as an .rds, and then in a new R session, loading the .rds, and using that object with various functions. For instance: 

With ctStanPostPredict(), I get: Error in array(x, c(length(x), 1L), if (!is.null(names(x))) list(names(x),  : 
  'data' must be of a vector type, was 'NULL'

For reference, I'm using ctStanFit() and a ctModel type = "stanct". 

Is there a better way to save and re-use models, or is there a different problem I am missing. Sorry in advance for missing something obvious. Happy to send along code or further information.

**Comment by cdriveraus** [comment]
Thanks Ross! You have actually pointed me to a bug in ctStanPostPredict that was occurring when nsamples requested is greater than the number of samples available -- this works fine on an optimized object because we can easily draw more samples but not for a fit sampled via stan. This is fixed on github now and it just samples the parameter posterior with replacement in order to generate the requested number of samples of new data. But it sounds like you are experiencing other issues with saved objects? In general saved objects and rstan are a bit of a headache but I thought I'd addressed those, so if you can point to where the issues are it will help a lot.

**Comment by Rjacobucci** [comment]
Thanks for your response. I'm still getting the same error -- can you confirm with what version of the package it is fixed in? I downloaded and installed 3.7.3 after your response. 

Additionally, I've tested:
- ctAddSamples(): I get "'data' must be of a vector type, was 'NULL'"
- ctKalman():  I get "Error : Exception: variable does not exist;" 

Ultimately this might be nullified by using the optimization version, but I'm still testing optimization vs. sampling on my data to see what seems to be the most accurate. The eventual goal is to train a model on a large dataset, but update the model as new data comes in. 

Thanks!

**Reply by cdriveraus** [reply]
Fixed in current github, 3.7.3. 

ctAddSamples is only available for optimized objects -- this is specified in the help file and I *had* coded a check on this but it was too far in to the function :) This is fixed. 

Re ctKalman -- when I fit a model using sampling, save, restart rstudio, load ctsem and load the saved fit, ctKalman works fine on this object. On ubuntu and windows at least. Can you confirm the issue using this minimal example? Bit confused why the fix isn't working for you and you're getting these issues... can you double check the version you have? Installation will fail if you try to install ctsem while it's already loaded, can you install from github in a fresh r session without any other r sessions running?

```
library(ctsem)

checkm<-ctModel(
  type='stanct',
  LAMBDA=diag(2))  

fit<-ctStanFit(ctsem::ctstantestdat,checkm,cores=2,control=list(max_treedepth=2),iter=100, optimize = F,nopriors=FALSE)

ctKalman(fit)
save(fit,file='cts.rda')


# restart -----------------------------------------------------------------

library(ctsem)
load("~/ctsem/cts.rda")
ctKalman(fit,plot=T)
ctStanPostPredict(fit)


```

**Comment by Rjacobucci** [comment]
Thanks again for your help! 

I can confirm that that code works for me, and I got it to work with my data. Part of the issue seemed to be that I was using a .rds file not .rda.

I'm still getting an issue with ctKalman: "Error : Exception: variable does not exist; processing stage=data initialization; variable name=tipredsdata; base type=matrix_d  (in 'model_ctsm' at line 307)"

I'm able to run the model, save it, load it, use ctStanFitUpdate, and neither the updated or original object works with ctKalman. If not straightforward to see a solution, I could send along some data and code if needed. At the least, here is the code I'm using up to that point:

mod2 <- ctModel(type="stanct",id="subject_id",time="time_0",
                n.latent=1,n.manifest=1,n.TDpred=3,#T0MEANS=0,
                TDpredNames=c("LIKE_I_DO_NOT_BELONG","LIKE_A_BURDEN","passive"),#,"passive"),
                manifestNames=c("active"),
                latentNames=c("active"),
                LAMBDA=1)
mod2$manifesttype = c(1)

fit3 <- ctStanFit(datalong=sub_dat,ctstanmodel=mod2,iter=300,chains=12,cores=12,optimcontrol=list(deoptim=T),
                  stationary=F,optimize=F,savesubjectmatrices=T,#nlcontrol=list(maxtimestep=10),
                  control=list(max_treedepth=20,adapt_delta=0.95))



save(fit3,file='rema_ctsem_crc1.rda')


fit22 = ctStanFitUpdate(fit3, data = sub_dat,refit=T,iter=20,recompile=T)


est_23329_1 = ctKalman(fit22,plot=TRUE) #

**Reply by cdriveraus** [reply]
.rds caused problems but .rda was ok? Interactions with stan do turn up some weird things occasionally...

I'm able to run that code, save out the fit, restart, load the fit, and run ctKalman on it. I'm also able to use ctStanFitUpdate (but recompile had to be set FALSE because I'm testing on windows with r 4.2 and CRAN version of rstan, which for the time being can't compile) on the loaded fit and then can still run ctKalman. What OS are you on? Why are you using ctStanFitUpdate? Is sub_dat really the same when you first fit and then when you run ctStanFitUpdate? ctStanFitUpdate is probably not the most robust of functions, it should work well for a) upgrading a model object to a newer version of ctsem (unless the internal model structure changed between versions, which is rare), or b) changing data, so long as the model parameters are the same. This latter point is probably where things are going wrong if your data to update is indeed different. I'm updating the help file now with these details :) You seem to be diving into the depths of ctsem functions, if you want to do a quick call to discuss what you're trying to do and where it's going wrong just let me know. Also, if you are using stan's sampler and have a binary observable, you 'may' want to set intoverstates=FALSE , as by default an approximation is used that allows optimized fits with binary data.
